### DISCUSSION TITLE
Language Support for Combining Expressions?

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5189] [Q&A] [answered]
Maybe I'm missing something but it seems like it should be really easy to combine expressions or use them as sub expressions.

How would I do this:
```
public task RemapUsers(Expression<Func<User, string>> expression){
  var Data = DbContext.Set<User>().Where(x => expression(x) == "Hello").ToLiust();
}
```

**Comment by jnm2** [comment]
I haven't tried this, but it seems like it could work:

```cs
public Task RemapUsers(Expression<Func<User, string>> expression)
{
    var predicate = Expression.Lambda<Func<User, bool>>(
        Expression.Equal(expression.Body, Expression.Constant("Hello")),
        expression.Parameters);

    var Data = DbContext.Set<User>().Where(predicate).ToList();
}
```

The general case of inlining an expression like this seems like it involves creating a new tree by rewriting the `expression.Body` to replace the `expression.Parameters` and building up the expression around it using the `Expression` factory API.

**Reply by TonyValenti** [reply]
Any chance the language could do all this for me?  This seems extremely complex.

**Reply by jnm2** [reply]
Sure seems handy occasionally. Maybe a complex proposal to give invocation expressions a new kind of meaning within expression trees when the receiver type is a lambda Expression instance.

The compiler would probably want a new Expression API to perform the inlining, probably, since it must happen at runtime. Unless the Expression API itself already supports expressing a call to an independently created lambda Expression instance.

**Reply by HaloFour** [reply]
I've had many cases when working with EF where I would have liked to have been able to write a snippet of expression tree that could be reused in other expression trees.  I did have a few helper functions which could combine expression trees or visit/replace method invocation expressions with a constant value or a chunk of expression tree, but it was awkward to use.

**Reply by canton7** [reply]
> The general case of inlining an expression like this seems like it involves creating a new tree by rewriting the expression.Body to replace the expression.Parameters

That's exactly it, see e.g. https://stackoverflow.com/a/58630886/1086121 for a general solution.

**Comment by svick** [comment, marked as answer]
You can use [LINQKit](https://github.com/scottksmith95/LINQKit) for this:

```c#
var Data = DbContext.Set<User>().AsExpandable().Where(x => expression.Invoke(x) == "Hello").ToList();
```

Considering that LINQKit exists and that this is a niche use case, I don't think it makes sense to add language support for this.
